heartbeat/security.go

```diff
@@ -0,0 +1,236 @@
+	// and add capabilities to make this actually run as a regular user. This also helps node in synthetics, which
+	// does not want to run as root. It's also just generally more secure.
+	if localUserName := os.Getenv("BEAT_SETUID_AS"); localUserName != "" && syscall.Geteuid() == 0 {
+		err := changeUser(localUserName)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// Attempt to set capabilities before we setup seccomp rules
+	// Note that we discard any errors because they are not actionable.
+	// The beat should use `getcap` at a later point to examine available capabilities
+	// rather than relying on errors from `setcap`
+	setCapabilities()
+
+	switch runtime.GOARCH {
+	case "amd64", "386":
+		err := setSeccompRules()
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func changeUser(localUserName string) error {
+	localUser, err := user.Lookup(localUserName)
+	if err != nil {
+		return fmt.Errorf("could not lookup '%s': %w", localUser, err)
+	}
+	localUserUid, err := strconv.Atoi(localUser.Uid)
+	if err != nil {
+		return fmt.Errorf("could not parse UID '%s' as int: %w", localUser.Uid, err)
+	}
+	localUserGid, err := strconv.Atoi(localUser.Gid)
+	if err != nil {
+		return fmt.Errorf("could not parse GID '%s' as int: %w", localUser.Uid, err)
+	}
+	// We include the root group because the docker image contains many directories (data,logs)
+	// that are owned by root:root with 0775 perms. The heartbeat user is in both groups
+	// in the container, but we need to repeat that here.
+	err = syscall.Setgroups([]int{localUserGid, 0})
+	if err != nil {
+		return fmt.Errorf("could not prsetgroups: %w", err)
+	}
+
+	// Set the main group as localUserUid so new files created are owned by the user's group
+	err = syscall.Setgid(localUserGid)
+	if err != nil {
+		return fmt.Errorf("could not set gid to %d: %w", localUserGid, err)
+	}
+
+	// Note this is not the regular SetUID! Look at the 'cap' package docs for it, it preserves
+	// capabilities post-SetUID, which we use to lock things down immediately
+	err = cap.SetUID(localUserUid)
+	if err != nil {
+		return fmt.Errorf("could not setuid to %d: %w", localUserUid, err)
+	}
+
+	// This may not be necessary, but is good hygeine, we do some shelling out to node/npm etc.
+	// and $HOME should reflect the user's preferences
+	return os.Setenv("HOME", localUser.HomeDir)
+}
+
+func setCapabilities() error {
+	// Start with an empty capability set
+	newcaps := cap.NewSet()
+	// Both permitted and effective are required! Permitted makes the permmission
+	// possible to get, effective makes it 'active'
+	err := newcaps.SetFlag(cap.Permitted, true, cap.NET_RAW)
+	if err != nil {
+		return fmt.Errorf("error setting permitted setcap: %w", err)
+	}
+	err = newcaps.SetFlag(cap.Effective, true, cap.NET_RAW)
+	if err != nil {
+		return fmt.Errorf("error setting effective setcap: %w", err)
+	}
+
+	// We do not want these capabilities to be inherited by subprocesses
```

Comment:
I believe this is for Node.js ?

Reply:
Yes, node doesn't need to do raw network stuff :)